pkg/plugin/pause.go

```diff
@@ -0,0 +1,147 @@
+	%[1]s pause foo
+`
+)
+
+// PauseOptions provides information required to manage ExtendedDaemonSet
+type PauseOptions struct {
+	configFlags *genericclioptions.ConfigFlags
+	args        []string
+
+	client client.Client
+
+	genericclioptions.IOStreams
+
+	userNamespace             string
+	userExtendedDaemonSetName string
+}
+
+// NewPauseOptions provides an instance of GetOptions with default values
+func NewPauseOptions(streams genericclioptions.IOStreams) *PauseOptions {
+	return &PauseOptions{
+		configFlags: genericclioptions.NewConfigFlags(false),
+
+		IOStreams: streams,
+	}
+}
+
+// NewCmdPause provides a cobra command wrapping PauseOptions
+func NewCmdPause(streams genericclioptions.IOStreams) *cobra.Command {
+	o := NewPauseOptions(streams)
+
+	cmd := &cobra.Command{
+		Use:          "pause an ExtendedDaemonSet canary replicaset",
+		Short:        "pause canary replicaset",
+		Example:      fmt.Sprintf(pauseExample, "kubectl"),
+		SilenceUsage: true,
+		RunE: func(c *cobra.Command, args []string) error {
+			if err := o.Complete(c, args); err != nil {
+				return err
+			}
+			if err := o.Validate(); err != nil {
+				return err
+			}
+			return o.Run()
+		},
+	}
+
+	o.configFlags.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+// Complete sets all information required for processing the command
+func (o *PauseOptions) Complete(cmd *cobra.Command, args []string) error {
+	o.args = args
+	var err error
+
+	clientConfig := o.configFlags.ToRawKubeConfigLoader()
+	// Create the Client for Read/Write operations.
+	o.client, err = NewClient(clientConfig)
+	if err != nil {
+		return fmt.Errorf("unable to instantiate client, err: %v", err)
+	}
+
+	o.userNamespace, _, err = clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	ns, err2 := cmd.Flags().GetString("namespace")
+	if err2 != nil {
+		return err
+	}
+	if ns != "" {
+		o.userNamespace = ns
+	}
+
+	if len(args) > 0 {
+		o.userExtendedDaemonSetName = args[0]
+	}
+
+	return nil
+}
+
+// Validate ensures that all required arguments and flag values are provided
+func (o *PauseOptions) Validate() error {
+
+	if len(o.args) < 1 {
+		return fmt.Errorf("the extendeddaemonset name is required")
+	}
+
+	return nil
+}
+
+// Run use to run the command
+func (o *PauseOptions) Run() error {
+	eds := &v1alpha1.ExtendedDaemonSet{}
+	err := o.client.Get(context.TODO(), client.ObjectKey{Namespace: o.userNamespace, Name: o.userExtendedDaemonSetName}, eds)
+	if err != nil && errors.IsNotFound(err) {
+		return fmt.Errorf("ExtendedDaemonSet %s/%s not found", o.userNamespace, o.userExtendedDaemonSetName)
+	} else if err != nil {
+		return fmt.Errorf("unable to get ExtendedDaemonSet, err: %v", err)
+	}
+
+	if eds.Spec.Strategy.Canary == nil {
+		return fmt.Errorf("the ExtendedDaemonset does not have a canary")
+	}
+
+	if eds.Spec.Strategy.Canary.Paused {
+		return fmt.Errorf("ExtendedDaemonset '%s/%s' deployment already paused", o.userNamespace, o.userExtendedDaemonSetName)
+	}
```

Comment:
Can we implement an unpause command ?

I'm thinking we can refactor a bit what we have currently and implement a new command `canary` with subcommands `validate` `pause` and `unpause`. I know this is extra work and wasn't in the scope of the task though. We can do it later.
cc @clamoriniere in case you have something to add here

Reply:
I don't mind expanding the scope here, it's a relatively small change. I'll add the subcommand.

Reply:
thank you 🙇

Reply:
here is an example on how we handle subcommands in the [operator kubectl-plugin ](https://github.com/DataDog/datadog-operator/blob/master/cmd/kubectl-datadog/agent/agent/agent.go#L37-L39)

Reply:
also [here](https://github.com/DataDog/datadog-operator/blob/master/cmd/kubectl-datadog/datadog/datadog.go#L39-L48)